File: pybit/javabot/utils/indicators.js

// --- Numeric Helpers ---
/** Safely formats a number to a fixed number of decimal places, returning 'N/A' for invalid numbers. */
function nfmt(n, digits = 4) {
    return Number.isFinite(n) ? n.toFixed(digits) : 'N/A';
}

/** Calculates Simple Moving Average for an array of numbers. */
function sma(values, period) {
    if (!Array.isArray(values) || values.length < period) return null;
    const slice = values.slice(-period);
    const sum = slice.reduce((a, b) => a + b, 0);
    return sum / period;
}

/** Calculates Exponential Moving Average for an array of numbers. */
function ema(values, period) {
    if (!Array.isArray(values) || values.length < period) return null;
    const k = 2 / (period + 1);
    let prev = sma(values.slice(0, period), period);
    if (prev == null) return null;
    for (let i = period; i < values.length; i++) {
        prev = (values[i] - prev) * k + prev;
    }
    return prev;
}

// --- Indicator Calculation Functions ---

/** Calculates SMA using numeric array helper. */
function calculateSMA(data, period, priceType = 'close') {
    if (!data || data.length < period) return null;
    const values = data.slice(-period).map(d => Number(d[priceType]));
    return sma(values, period);
}

/** Calculates EMA using numeric array helper. */
function calculateEMA(data, period, priceType = 'close') {
    if (!data || data.length < period) return null;
    const values = data.map(d => Number(d[priceType]));
    return ema(values, period);
}

/** Calculates MACD with optimized EMA series and signal line calculation. */
function calculateMACD(data, fastPeriod, slowPeriod, signalPeriod) {
    if (!data || data.length < slowPeriod + signalPeriod - 1) return null;
    const closes = data.map(d => Number(d.close));

    const emaSeries = (period) => {
        if (!closes || closes.length < period) return null;
        const out = new Array(closes.length).fill(null);
        let prev = sma(closes.slice(0, period), period);
        if (prev == null) return null;
        out[period - 1] = prev;
        const k = 2 / (period + 1);
        for (let i = period; i < closes.length; i++) {
            prev = (closes[i] - prev) * k + prev;
            out[i] = prev;
        }
        return out;
    };

    const emaFast = emaSeries(fastPeriod);
    const emaSlow = emaSeries(slowPeriod);
    if (!emaFast || !emaSlow) return null;

    const macd = closes.map((_, i) => (emaFast[i] != null && emaSlow[i] != null) ? (emaFast[i] - emaSlow[i]) : null);
    const macdVals = macd.filter(v => v != null);
    if (macdVals.length < signalPeriod) return null;

    const signalAll = new Array(macd.length).fill(null);
    let prevSignal = sma(macdVals.slice(0, signalPeriod), signalPeriod);
    if (prevSignal == null) return null;

    let firstSignalAt = macd.findIndex(v => v != null) + signalPeriod - 1;
    if (firstSignalAt >= macd.length) return null;
    signalAll[firstSignalAt] = prevSignal;

    const k = 2 / (signalPeriod + 1);
    for (let i = firstSignalAt + 1; i < macd.length; i++) {
        if (macd[i] == null) continue;
        prevSignal = (macd[i] - prevSignal) * k + prevSignal;
        signalAll[i] = prevSignal;
    }

    const last = macd.length - 1;
    if (macd[last] == null || signalAll[last] == null) return null;
    return {
        macdLine: macd[last],
        signalLine: signalAll[last],
        histogram: macd[last] - signalAll[last],
    };
}

/** Calculates RSI with Wilder's smoothing. */
function calculateRSI(data, period) {
    if (!data || data.length < period + 1) return null;
    const closes = data.map(d => Number(d.close));
    let gain = 0, loss = 0;

    for (let i = 1; i <= period; i++) {
        const change = closes[closes.length - i] - closes[closes.length - i - 1];
        if (change > 0) gain += change; else loss -= change;
    }
    let avgGain = gain / period;
    let avgLoss = loss / period;

    for (let i = period + 1; i < closes.length; i++) {
        const change = closes[closes.length - i] - closes[closes.length - i - 1];
        const g = change > 0 ? change : 0;
        const l = change < 0 ? -change : 0;
        avgGain = (avgGain * (period - 1) + g) / period;
        avgLoss = (avgLoss * (period - 1) + l) / period;
    }

    if (avgLoss === 0) return 100;
    const rs = avgGain / avgLoss;
    return 100 - 100 / (1 + rs);
}

/** Calculates Stochastic Oscillator (%K and %D). */
function calculateStochastic(data, kPeriod, dPeriod) {
    if (!data || data.length < kPeriod || data.length < dPeriod) return null;
    const prices = data.map(d => Number(d.close));
    const highs = data.map(d => Number(d.high));
    const lows = data.map(d => Number(d.low));

    const relevantPricesK = prices.slice(prices.length - kPeriod);
    const relevantHighsK = highs.slice(prices.length - kPeriod);
    const relevantLowsK = lows.slice(lows.length - kPeriod);

    const highestHighK = Math.max(...relevantHighsK);
    const lowestLowK = Math.min(...relevantLowsK);
    const currentCloseK = relevantPricesK[relevantPricesK.length - 1];

    let kValue = 0;
    if (highestHighK !== lowestLowK) {
        kValue = ((currentCloseK - lowestLowK) / (highestHighK - lowestLowK)) * 100;
    } else {
        kValue = 50;
    }

    const kValues = [];
    for (let i = 0; i < prices.length; i++) {
        const currentSlice = data.slice(0, i + 1);
        if (currentSlice.length >= kPeriod) {
            const currentPrices = currentSlice.map(d => Number(d.close));
            const currentHighs = currentSlice.map(d => Number(d.high));
            const currentLows = currentSlice.map(d => Number(d.low));

            const currentHighestHigh = Math.max(...currentHighs.slice(-kPeriod));
            const currentLowestLow = Math.min(...currentLows.slice(-kPeriod));
            const currentClose = currentPrices[currentPrices.length - 1];

            let currentK = 0;
            if (currentHighestHigh !== currentLowestLow) {
                currentK = ((currentClose - currentLowestLow) / (currentHighestHigh - currentLowestLow)) * 100;
            } else {
                currentK = 50;
            }
            kValues.push(currentK);
        } else {
            kValues.push(null);
        }
    }
    const validKValues = kValues.filter(val => val !== null);
    const dValue = sma(validKValues, dPeriod);

    return { k: kValue, d: dValue };
}

/** Calculates ATR. */
function calculateATR(data, period) {
    if (!data || data.length < period + 1) return null;
    const highs = data.map(d => Number(d.high));
    const lows = data.map(d => Number(d.low));
    const closes = data.map(d => Number(d.close));

    const trueRanges = [];
    for (let i = 1; i < highs.length; i++) {
        const tr1 = highs[i] - lows[i];
        const tr2 = Math.abs(highs[i] - closes[i - 1]);
        const tr3 = Math.abs(lows[i] - closes[i - 1]);
        trueRanges.push(Math.max(tr1, tr2, tr3));
    }

    const initialATR = sma(trueRanges.slice(0, period), period);
    if (initialATR === null) return null;

    let atr = initialATR;
    const multiplier = 1 / period;

    for (let i = period; i < trueRanges.length; i++) {
        atr = (trueRanges[i] - atr) * multiplier + atr;
    }
    return atr;
}

/** Calculates Bollinger Bands. */
function calculateBollingerBands(data, period, stdDev) {
    if (!data || data.length < period) return null;
    const prices = data.map(d => Number(d.close));

    const middleBand = calculateSMA(data, period, 'close');
    if (middleBand === null) return null;

    const relevantPrices = prices.slice(prices.length - period);
    const mean = middleBand;
    const squaredDifferences = relevantPrices.map(p => Math.pow(p - mean, 2));
    const variance = squaredDifferences.reduce((a, b) => a + b, 0) / period;
    const standardDeviation = Math.sqrt(variance);

    const upperBand = middleBand + (standardDeviation * stdDev);
    const lowerBand = middleBand - (standardDeviation * stdDev);

    return { middleBand, upperBand, lowerBand };
}

/** Calculates OBV. */
function calculateOBV(data) {
    if (!data || data.length < 2) return null;
    const prices = data.map(d => Number(d.close));
    const volumes = data.map(d => Number(d.volume));

    let obv = 0;
    for (let i = 0; i < prices.length; i++) {
        if (i === 0) {
            obv = volumes[i];
        } else {
            if (prices[i] > prices[i - 1]) {
                obv += volumes[i];
            } else if (prices[i] < prices[i - 1]) {
                obv -= volumes[i];
            }
        }
    }
    return obv;
}

/** Calculates Williams %R. */
function calculateWilliamsR(data, period) {
    if (!data || data.length < period) return null;
    const prices = data.map(d => Number(d.close));
    const highs = data.map(d => Number(d.high));
    const lows = data.map(d => Number(d.low));

    const relevantPrices = prices.slice(prices.length - period);
    const relevantHighs = highs.slice(prices.length - period);
    const relevantLows = lows.slice(prices.length - period);

    const highestHigh = Math.max(...relevantHighs);
    const lowestLow = Math.min(...relevantLows);
    const currentClose = relevantPrices[relevantPrices.length - 1];

    if (highestHigh === lowestLow) return 0;

    const williamsR = ((highestHigh - currentClose) / (highestHigh - lowestLow)) * -100;
    return williamsR;
}

/** Calculates CMF using the last N bars. */
function calculateCMF(data, period) {
    if (!data || data.length < period) return null;
    const slice = data.slice(-period);
    let mfvSum = 0, volSum = 0;

    for (const d of slice) {
        const high = Number(d.high), low = Number(d.low), close = Number(d.close), vol = Number(d.volume);
        const range = high - low;
        if (range === 0 || !Number.isFinite(range) || !Number.isFinite(vol)) continue;

        const mfMultiplier = ((close - low) - (high - close)) / range;
        const mfv = mfMultiplier * vol;
        mfvSum += mfv;
        volSum += vol;
    }

    if (volSum === 0) return 0;
    return mfvSum / volSum;
}

/** Calculates Elder-Ray. */
function calculateElderRay(data, period) {
    if (!data || data.length < period) return null;
    const highs = data.map(d => Number(d.high));
    const lows = data.map(d => Number(d.low));

    const ema = calculateEMA(data, period, 'close');
    if (ema === null) return null;

    const currentHigh = highs[highs.length - 1];
    const currentLow = lows[lows.length - 1];

    const bullishPower = currentHigh - ema;
    const bearishPower = currentLow - ema;

    return { bullishPower, bearishPower };
}

/** Calculates Keltner Channels. */
function calculateKeltnerChannels(data, period, atrMultiplier) {
    if (!data || data.length < period) return null;
    const middleBand = calculateEMA(data, period, 'close');
    if (middleBand === null) return null;

    const atr = calculateATR(data, period);
    if (atr === null) return null;

    const upperChannel = middleBand + (atr * atrMultiplier);
    const lowerChannel = middleBand - (atr * atrMultiplier);

    return { middleBand, upperChannel, lowerChannel };
}

/** Calculates Aroon Indicator (fixed recency). */
function calculateAroon(data, period) {
    if (!data || data.length < period) return null;
    const highs = data.map(d => Number(d.high));
    const lows = data.map(d => Number(d.low));

    const windowHighs = highs.slice(-period);
    const windowLows = lows.slice(-period);

    const highestHigh = Math.max(...windowHighs);
    const lowestLow = Math.min(...windowLows);

    const hhIdx = windowHighs.lastIndexOf(highestHigh);
    const llIdx = windowLows.lastIndexOf(lowestLow);

    const periodsSinceHH = (period - 1) - hhIdx;
    const periodsSinceLL = (period - 1) - llIdx;

    const aroonUp = ((period - periodsSinceHH) / period) * 100;
    const aroonDown = ((period - periodsSinceLL) / period) * 100;

    return { aroonUp, aroonDown };
}

module.exports = {
    nfmt,
    calculateSMA,
    calculateEMA,
    calculateMACD,
    calculateRSI,
    calculateStochastic,
    calculateATR,
    calculateBollingerBands,
    calculateOBV,
    calculateWilliamsR,
    calculateCMF,
    calculateElderRay,
    calculateKeltnerChannels,
    calculateAroon
};

File: html/indicators.py
import math

# --- Numerically Stable Helper Functions ---


def _calculate_wilder_rma(data: list[float], period: int) -> list[float]:
    """Calculates Wilder's Recursive Moving Average (RMA)."""
    if not data or len(data) < period:
        return [0.0] * len(data)

    rma_values = [0.0] * len(data)
    alpha = 1.0 / period

    # Initial SMA
    rma_values[period - 1] = sum(data[:period]) / period

    # Recursive calculation
    for i in range(period, len(data)):
        rma_values[i] = alpha * data[i] + (1 - alpha) * rma_values[i - 1]

    return rma_values


def _calculate_ema(data: list[float], period: int) -> list[float]:
    """Calculates Exponential Moving Average (EMA)."""
    if not data or len(data) < period:
        return [0.0] * len(data)

    ema_values = [0.0] * len(data)
    smoothing_factor = 2.0 / (period + 1)

    # Initial SMA
    ema_values[period - 1] = sum(data[:period]) / period

    # Exponential calculation
    for i in range(period, len(data)):
        ema_values[i] = (data[i] - ema_values[i - 1]) * smoothing_factor + ema_values[
            i - 1
        ]

    return ema_values


def _calculate_sma(data: list[float], period: int) -> list[float]:
    """Calculates Simple Moving Average (SMA)."""
    if not data or len(data) < period:
        return [0.0] * len(data)

    sma_values = [0.0] * len(data)
    current_sum = sum(data[:period])
    sma_values[period - 1] = current_sum / period

    for i in range(period, len(data)):
        current_sum += data[i] - data[i - period]
        sma_values[i] = current_sum / period

    return sma_values


# --- Indicator Calculation Functions ---


def _calculate_atr(klines: list[dict], period: int) -> list[float]:
    """Calculates Average True Range (ATR)."""
    trs = [0.0] * len(klines)
    for i in range(1, len(klines)):
        kline = klines[i]
        prev_kline = klines[i - 1]
        tr = max(
            kline["high"] - kline["low"],
            abs(kline["high"] - prev_kline["close"]),
            abs(kline["low"] - prev_kline["close"]),
        )
        trs[i] = tr
    return _calculate_wilder_rma(trs, period)


def _calculate_supertrend(klines: list[dict], length: int, multiplier: float) -> dict:
    """Calculates Supertrend indicator."""
    closes = [k["close"] for k in klines]
    atr_values = _calculate_atr(klines, length)

    upper_band = [0.0] * len(klines)
    lower_band = [0.0] * len(klines)
    supertrend = [0.0] * len(klines)
    direction = [1] * len(klines)

    for i in range(length, len(klines)):
        hl2 = (klines[i]["high"] + klines[i]["low"]) / 2
        upper_band[i] = hl2 + multiplier * atr_values[i]
        lower_band[i] = hl2 - multiplier * atr_values[i]

        # Adjust bands based on previous close
        if closes[i - 1] > upper_band[i - 1]:
            upper_band[i] = min(upper_band[i], upper_band[i - 1])
        if closes[i - 1] < lower_band[i - 1]:
            lower_band[i] = max(lower_band[i], lower_band[i - 1])

        # Determine direction and final supertrend value
        if closes[i] > upper_band[i - 1]:
            direction[i] = 1
        elif closes[i] < lower_band[i - 1]:
            direction[i] = -1
        else:
            direction[i] = direction[i - 1]

        supertrend[i] = lower_band[i] if direction[i] == 1 else upper_band[i]

    return {"supertrend": supertrend[-1], "direction": direction[-1]}


def _calculate_rsi(closes: list[float], length: int) -> float:
    """Calculates Relative Strength Index (RSI) using Wilder's smoothing."""
    deltas = [closes[i] - closes[i - 1] for i in range(1, len(closes))]
    gains = [d if d > 0 else 0.0 for d in deltas]
    losses = [-d if d < 0 else 0.0 for d in deltas]

    avg_gain = _calculate_wilder_rma(gains, length)[-1]
    avg_loss = _calculate_wilder_rma(losses, length)[-1]

    if avg_loss == 0:
        return 100.0

    rs = avg_gain / avg_loss
    return 100.0 - (100.0 / (1.0 + rs))


def _calculate_ehlers_fisher(klines: list[dict], period: int) -> float:
    """Calculates Ehlers-Fisher Transform."""
    values = [0.0] * len(klines)
    fishers = [0.0] * len(klines)

    for i in range(period, len(klines)):
        period_klines = klines[i - period + 1 : i + 1]
        hh = max(k["high"] for k in period_klines)
        ll = min(k["low"] for k in period_klines)

        price = (klines[i]["high"] + klines[i]["low"]) / 2

        # Normalize price from -1 to 1
        val = 0.0
        if hh - ll != 0:
            val = 2 * ((price - ll) / (hh - ll)) - 1

        # Apply smoothing
        values[i] = 0.33 * val + 0.67 * values[i - 1]

        # Clamp values to avoid math domain errors
        clamped_val = max(min(values[i], 0.999), -0.999)

        # Fisher Transform
        fishers[i] = (
            0.5 * math.log((1 + clamped_val) / (1 - clamped_val)) + 0.5 * fishers[i - 1]
        )

    return fishers[-1]


def _calculate_macd(
    closes: list[float], fast_period: int, slow_period: int, signal_period: int,
) -> dict:
    """Calculates Moving Average Convergence Divergence (MACD)."""
    fast_ema = _calculate_ema(closes, fast_period)
    slow_ema = _calculate_ema(closes, slow_period)

    macd_line = [f - s for f, s in zip(fast_ema, slow_ema, strict=False)]
    signal_line = _calculate_ema(macd_line, signal_period)
    histogram = [m - s for m, s in zip(macd_line, signal_line, strict=False)]

    return {
        "macd_line": macd_line[-1],
        "signal_line": signal_line[-1],
        "histogram": histogram[-1],
    }


def _calculate_bollinger_bands(
    closes: list[float], period: int, std_dev: float,
) -> dict:
    """Calculates Bollinger Bands."""
    middle_band = _calculate_sma(closes, period)

    # Calculate standard deviation
    std_dev_values = [0.0] * len(closes)
    for i in range(period - 1, len(closes)):
        window = closes[i - period + 1 : i + 1]
        mean = middle_band[i]
        variance = sum([(x - mean) ** 2 for x in window]) / period
        std_dev_values[i] = math.sqrt(variance)

    upper_band = [
        m + (s * std_dev) for m, s in zip(middle_band, std_dev_values, strict=False)
    ]
    lower_band = [
        m - (s * std_dev) for m, s in zip(middle_band, std_dev_values, strict=False)
    ]

    return {
        "middle_band": middle_band[-1],
        "upper_band": upper_band[-1],
        "lower_band": lower_band[-1],
    }


# --- Main Public Function ---


def calculate_indicators(klines: list[dict], config: dict) -> dict | None:
    """Calculates all trading indicators from a list of kline data.
    Returns a dictionary of indicator values or None if data is insufficient.
    """
    max_lookback = max(
        config.get("supertrend_length", 10),
        config.get("rsi_length", 14) + 1,  # RSI needs one more for deltas
        config.get("ef_period", 10),
        config.get("macd_slow_period", 26),
        config.get("bb_period", 20),
    )

    if not klines or len(klines) < max_lookback:
        return None

    closes = [float(k["close"]) for k in klines]

    return {
        "supertrend": _calculate_supertrend(
            klines, config["supertrend_length"], config["supertrend_multiplier"],
        ),
        "rsi": _calculate_rsi(closes, config["rsi_length"]),
        "fisher": _calculate_ehlers_fisher(klines, config["ef_period"]),
        "macd": _calculate_macd(
            closes,
            config["macd_fast_period"],
            config["macd_slow_period"],
            config["macd_signal_period"],
        ),
        "bollinger_bands": _calculate_bollinger_bands(
            closes, config["bb_period"], config["bb_std_dev"],
        ),
    }
